pkg/controllers/cluster/actions/upgrade_version.go

```diff
@@ -3,54 +3,966 @@ package actions
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/retry"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
 const (
 	ClusterVersionUpgradeAction = "rack-version-upgrade"
+	actionTimeout               = time.Minute
 )
 
 type ClusterVersionUpgrade struct {
-	Cluster *scyllav1alpha1.ScyllaCluster
-	logger  log.Logger
+	Cluster        *scyllav1alpha1.ScyllaCluster
+	ScyllaClient   *scyllaclient.Client
+	ClusterSession cqlSession
+
+	ipMapping map[string]string
+
+	currentRack *scyllav1alpha1.RackSpec
+	currentNode *corev1.Pod
+
+	cc         client.Client
+	kubeClient kubernetes.Interface
+	recorder   record.EventRecorder
+	logger     log.Logger
 }
 
 func NewClusterVersionUpgradeAction(c *scyllav1alpha1.ScyllaCluster, l log.Logger) *ClusterVersionUpgrade {
 	return &ClusterVersionUpgrade{
-		Cluster: c,
-		logger:  l,
+		Cluster:   c,
+		logger:    l,
+		ipMapping: map[string]string{},
 	}
 }
 
 func (a *ClusterVersionUpgrade) Name() string {
 	return ClusterVersionUpgradeAction
 }
 
+var ScyllaClientForClusterFunc = func(ctx context.Context, cc client.Client, hosts []string, logger log.Logger) (*scyllaclient.Client, error) {
+	cfg := scyllaclient.DefaultConfig(hosts...)
+	return scyllaclient.NewClient(cfg, logger)
+}
+
+type cqlSession interface {
+	AwaitSchemaAgreement(ctx context.Context) error
+}
+
+var NewSessionFunc = func(hosts []string) (cqlSession, error) {
+	cluster := gocql.NewCluster(hosts...)
+	return gocqlx.WrapSession(cluster.CreateSession())
+}
+
+func (a *ClusterVersionUpgrade) nonMaintenanceHosts(ctx context.Context) ([]string, error) {
+	var hosts []string
+	for _, r := range a.Cluster.Spec.Datacenter.Racks {
+		services, err := util.GetMemberServicesForRack(ctx, r, a.Cluster, a.cc)
+		if err != nil {
+			return nil, errors.Wrap(err, "get member services for rack")
+		}
+
+		for _, s := range services {
+			a.ipMapping[s.Name] = s.Spec.ClusterIP
+			if _, ok := s.Labels[naming.NodeMaintenanceLabel]; !ok {
+				hosts = append(hosts, s.Spec.ClusterIP)
+			}
+		}
+	}
+
+	return hosts, nil
+}
+
+type upgradeProcedure int
+
+const (
+	majorUpgradeProcedure = iota
+	patchUpgradeProcedure
+)
+
+func (a *ClusterVersionUpgrade) upgradeProcedure(ctx context.Context) upgradeProcedure {
+	if a.Cluster.Status.Upgrade != nil {
+		return majorUpgradeProcedure
+	}
+	var fromVersion string
+	// Fetch version from any rack status, they should be in the same version.
+	for _, r := range a.Cluster.Spec.Datacenter.Racks {
+		fromVersion = a.Cluster.Status.Racks[r.Name].Version
+		break
+	}
+
+	oldVersion, err := semver.Parse(fromVersion)
+	if err != nil {
+		a.logger.Info(ctx, "Invalid from semantic version", "version", fromVersion)
+		return majorUpgradeProcedure
+	}
+	newVersion, err := semver.Parse(a.Cluster.Spec.Version)
+	if err != nil {
+		a.logger.Info(ctx, "Invalid to semantic version", "version", fromVersion)
+		return majorUpgradeProcedure
```

Comment:
How this is supposed to work?

Reply:
I'll change name to "generic upgrade" procedure. For example upgrading to `nightly` shouldn't be rejected, and it would be rejected because it's not semantic, the same way Scylla versions aren't. Logic whether given version transition should be allowed or not, should be in webhook where change can be rejected with aproperiate message back to the user.